crates/commitlog/src/repo/mod.rs

```diff
@@ -0,0 +1,153 @@
+    /// [`io::ErrorKind::AlreadyExists`] if the segment already exists.
+    ///
+    /// It is permissible, however, to successfully return the new segment if
+    /// it is completely empty (i.e. [`create_segment_writer`] did not previously
+    /// succeed in writing the segment header).
+    fn create_segment(&self, offset: u64) -> io::Result<Self::Segment>;
+
+    /// Open an existing segment at the minimum transaction offset `offset`.
+    ///
+    /// Must return [`io::ErrorKind::NotFound`] if a segment with the given
+    /// `offset` does not exist.
+    ///
+    /// The method does not guarantee that the segment is non-empty -- this case
+    /// will be caught by [`open_segment_writer`] and [`open_segment_reader`]
+    /// respectively.
+    fn open_segment(&self, offset: u64) -> io::Result<Self::Segment>;
+
+    /// Remove the segment at the minimum transaction offset `offset`.
+    ///
+    /// Return [`io::ErrorKind::NotFound`] if no such segment exists.
+    fn remove_segment(&self, offset: u64) -> io::Result<()>;
+
+    /// Traverse all segments in this repository and return list of their
+    /// offsets, sorted in ascending order.
+    fn existing_offsets(&self) -> io::Result<Vec<u64>>;
+}
+
+/// Create a new segment [`Writer`] with `offset`.
+///
+/// Immediately attempts to write the segment header with the supplied
+/// `log_format_version`.
+///
+/// If the segment already exists, [`io::ErrorKind::AlreadyExists`] is returned.
+pub fn create_segment_writer<R: Repo>(repo: &R, opts: Options, offset: u64) -> io::Result<Writer<R::Segment>> {
+    let mut storage = repo.create_segment(offset)?;
+    Header {
+        log_format_version: opts.log_format_version,
+        checksum_algorithm: Commit::CHECKSUM_ALGORITHM,
+    }
+    .write(&mut storage)?;
+    storage.fsync()?;
+
+    Ok(Writer {
+        commit: Commit {
+            min_tx_offset: offset,
+            n: 0,
+            records: Vec::new(),
+        },
+        inner: io::BufWriter::new(storage),
+
+        min_tx_offset: offset,
+        bytes_written: Header::LEN as u64,
+
+        max_records_in_commit: opts.max_records_in_commit,
+    })
+}
+
+/// Open the existing segment at `offset` for writing.
+///
+/// This will traverse the segment in order to find the offset of the next
+/// commit to write to it, which may fail for various reasons.
+///
+/// If the traversal is successful, the segment header is checked against the
+/// `max_log_format_version`, and [`io::ErrorKind::InvalidData`] is returned if
+/// the segment's log format version is greater than the given value. Likewise
+/// if the checksum algorithm stored in the segment header cannot be handled
+/// by this crate.
+///
+/// If only a (non-empty) prefix of the segment could be read due to a failure
+/// to decode a [`Commit`], the segment [`Metadata`] read up to the faulty
+/// commit is returned in an `Err`. In this case, a new segment should be
+/// created for writing.
```

Comment:
Should the segment also be truncated? Due to the language in the proposal [here](https://github.com/clockworklabs/SpacetimeDBPrivate/blob/master/proposals/0003-commitlog-format.md#log-integrity)

Reply:
Good catch, I did indeed deviate from this approach. Probably should submit an amendment to the spec.

I realized that, unless `write(2)` can lie and return an error even though it has written everything, we can fairly easily skip over partial data (not least thanks to checksumming).

Because a new segment is started when a write error occurs, we'll retain data potentially useful for forensics in case of a bad disk.

Explicit truncation is still needed for consensus-based replication, though.

Reply:
Made an issue for amending the spec, no rush on that though.

Reply:
FWIW Kim now that I think about it

> unless write(2) can lie and return an error even though it has written everything

I think this may be possible. write(2) may return after data is written to a kernel buffer, but *before* it is actually written to the disk. Thus it would be possible to write(2) a commit, it return no error, only partially write that commit to disk, have a disk-write-error, and then have us write another commit to disk without us ever seeing an error.

Reply:
It's possible I'm missing some other context like, we check that all of the commit offsets are sequential or something (and if not we truncate)

Reply:
Sync'd with Kim. We are indeed checking for sequential commit numbers when reloading a database. Thus if we encounter this type of situation, it will require manual intervention to recover from (if that is even possible). This error must have occurred silently as illustrated in my scenario above, so there's nothing we could have really done to detect it when it happened anyway.

Could we expand the comment above to add the stipulation that this behavior of skipping corrupted messages is only valid if sequential ordering of commits is verified at startup, and provide my example as an illustration as why that is necessary.

Reply:
> but it can lie in this way:
> 
> > write(2) can return no error even though it _hasn't_ written everything

Wait... we cannot assume `write(2)` is behaving arbitrarily.
Namely, if we assume that write calls can be reordered arbitrarily, an append-only log is impossible.
We also must assume that `fsync(2)` won't return successfully if it hasn't flushed to disk.

What is indeed not correctly handled here is that an `fsync(2)` error should cause the active segment to be discarded. And because everything is happening asynchronously, failure to open a fresh segment must result in a panic (i.e. we must not keep writing to the old segment).

I'll make a new patch implementing this.

> Could we expand the comment above

I don't think a module which is not even exported from the crate is the right place to explain how different parts of the crate need to play together. Once we're overall confident with the implementation, I'd rather add some extensive crate-level docs, also summarizing the spec which is not accessible for everyone.

Reply:
This seems fine to me, as long as we don't forget to add that comment in at the end.

Reply:
The commentary states what happens, namely that one cannot resume a segment when it contains data which cannot be decoded as a commit. A new segment is created then, from the offset of the last good commit.

This can go wrong due to caching effects: re-reading from disk might actually be reading from the page cache. There is no remedy for this, but the next best thing is to bypass the page cache.

I don't know what else should be added here.